---
title: "Support for introgression of Sb using Twisst"
author: "Rodrigo Pracana, QMUL"
date: "07/08/2021"
output: html_document
---

```{r setup, include=FALSE, cache = FALSE}


make_sci <- function(x) {
    sprintf("%.1fx10^{%d}",
          x/10^floor(log10(abs(x))),
          floor(log10(abs(x))))
}

pretty_number_inner <- function(x) {
  if (is.na(x)) {
    x <- NA
  } else if (x == 0) {
    x <- as.character(x)
  } else if (x < 1e-4) {
    x <- as.character(make_sci(x))
  } else if  (x < 1e-2) {
     x <- as.character(round(x, 4))
  } else if (x < 1e4) {
     x <- as.character(round(x, 2))
  } else {
     x <- as.character(make_sci(x))
  }
  return(x)
}

pretty_numbers <- function(x) {
  sapply(x, pretty_number_inner)
}

knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)


library(tidyverse)
library(patchwork)
library(broom)

library(ape)
source("twisst/plot_twisst.R")

dir.create(recursive = T, "results/twisst_figures")
```

```{r}

# Correct function
#function for subsetting the twisst object by a set of topologies
subset.twisst.by.topos <- function(twisst_object, topos){
    l <- list()
    regions <- names(twisst_object$weights)
    l$window_data <- twisst_object$window_data
    l$n_regions <- twisst_object$n_regions
    l$lengths <- twisst_object$lengths
    l$pos <- twisst_object$pos
    l$weights_raw <- sapply(regions, function(region) twisst_object$weights_raw[[region]][,topos], simplify=F)
    l$weights <- sapply(regions, function(region) twisst_object$weights[[region]][,topos], simplify=F)
    l$weights_mean <- sapply(regions, function(region) twisst_object$weights_mean[[region]][topos], simplify=F)
    l$weights_overall_mean <- twisst_object$weights_overall_mean[topos]
    l$topos <- twisst_object$topos[topos]
    l
    }


#function for subsetting the twisst object by specific regions
subset.twisst.by.regions <- function(twisst_object, regions){
    l <- list()
    regions <- names(twisst_object$weights[regions])
    l$window_data <- twisst_object$window_data[regions]
    l$n_regions <- length(regions)
    l$lengths <- twisst_object$lengths[regions]
    l$pos <- twisst_object$pos[regions]
    l$weights_raw <- twisst_object$weights_raw[regions]
    l$weights <- twisst_object$weights[regions]
    l$weights_mean <- twisst_object$weights_mean[regions]
    l$weights_overall_mean <- apply(rbindlist(l$weights), 2, mean, na.rm=T)
    l$topos <- twisst_object$topos
    l
    }

```

```{r load_data, include=FALSE}


#weights file with a column for each topology
weights_file <- "results/2021-08-07-twisst/weights_output.csv.gz"

#coordinates file for each window
window_data_file <- "results/window_coordinates"

all_twisst <- import.twisst(weights_files = weights_file,
                            window_data_files = window_data_file)

# Table
full_coords <- read.table(window_data_file, header = T)

full_coords %>%
  pivot_longer(cols = !c(1:6), names_to = "window_gene",values_to = "gene_name") ->
  all_genes

all_genes <- unique(sort(all_genes$gene_name))

```

## Introduction

We have grouped the `r length(all_genes)` genes in the supergene in `r nrow(full_coords)` windows of five genes (sliding three by three). For each gene, we kept individuals with less than 5% of missing genotypes representing *S. invicta*, *S. richteri*, *S. saevissima*, *S. pusillignis* and *S. geminata*. We built a tree with RAxML for each of the windows. For each window, we used Twisst to find support for each of the `r length(all_twisst$topos)` possible topologies given the number of clades in our dataset. Samples carrying SB and Sb are each considered different clades for *S. invicta*, *S. richteri*.

## Support for each topology

By plotting the average weight of each topology across the windows, we show that a few topologies have much greater support than others.

```{r}

weights_mean <- data.frame(topology = names(all_twisst$weights_overall_mean),
                           mean = all_twisst$weights_overall_mean)


ord <- order(all_twisst$weights_overall_mean, decreasing=T)

N=length(ord)
barplot(all_twisst$weights_overall_mean[ord],
            xaxt="n", las=1, xlab = "Topology", ylab="Average weighting",
        space = 0.2, xlim = c(0.2, 1.2*N))


```

We first look at chromosome 1. If we take the 12 topologies with the highest average support across the `r nrow(full_coords)` windows, we show that the first four topologies have the greatest support.

```{r, fig.width=25}

#code for plotting a summary barplot
plot.twisst.summary <- function(twisst_object, order_by_weights=TRUE, only_best=NULL, cols=topo_cols,
                                x_scale=0.12, y_scale=0.15, direction="right", col="black", col.label="black",
                                label_offset = 0.1, lwd=NULL, cex=NULL){

    # Either order 1-15 or order with highest weigted topology first

    if (order_by_weights == TRUE) {
        ord <- order(twisst_object$weights_overall_mean, decreasing=T)
        if (is.null(only_best) == FALSE) ord=ord[1:only_best]
        }
    else ord <- 1:length(twisst_object$topos)

    N=length(ord)

    #set the plot layout, with the tree panel one third the height of the barplot panel
    layout(matrix(c(2,1)), heights=c(1,1))

    par(mar = c(1,4,0.5,1))

    #make the barplot
    x=barplot(twisst_object$weights_overall_mean[ord], col = cols[ord],
            xaxt="n", las=1, ylab="Average weighting", space = 0.2, xlim = c(0.2, 1.2*N))

    #draw the trees
    #first make an empty plot for the trees. Ensure left and right marhins are the same
    par(mar=c(0,4,0,1))
    plot(0,cex=0,xlim = c(0.2, 1.2*N), xaxt="n",yaxt="n",xlab="",ylab="",ylim=c(0,2.1), bty="n")

    #now run the draw.tree function for each topology. You can set x_scale and y_scale to alter the tree width and height.
    for (i in 1:length(ord)){
        draw.tree(twisst_object$topos[[ord[i]]], x=x[i]+.2, y=0, x_scale=x_scale, y_scale=y_scale,
                col=cols[ord[i]], label_offset=label_offset, cex=cex, lwd=lwd)
        }

    #add labels for each topology
    text(x,0,names(twisst_object$topos)[ord],col=cols[ord])
    }



# Subset to chr1
chr1_topo <- subset.twisst.by.regions(all_twisst, "chr1")

# Subset to the most abundant topologies
chr1_top12 <- order(chr1_topo$weights_overall_mean, decreasing=T)[1:12]

#subset twisst object for these
chr1_top12 <- subset.twisst.by.topos(chr1_topo, chr1_top12)


#this can then be used in all the same plotting functions above.
plot.twisst.summary(chr1_top12, lwd=3, cex=0.7)

pdf("results/2021-08-07-twisst/summary_top12_chr1.pdf", width=31, height=8)
plot.twisst.summary(chr1_top12, lwd=3, cex=0.7)
dev.off()

```

For chromosome 16, we get the same topology, as it includes many windows outside the supergene region:

```{r , fig.width=25}

# Subset to chr16
chr16_topo <- subset.twisst.by.regions(all_twisst, "chr16")

# Subset to the most abundant topologies
chr16_top12 <- order(chr16_topo$weights_overall_mean, decreasing=T)[1:12]

#subset twisst object for these
chr16_top12 <- subset.twisst.by.topos(chr16_topo, chr16_top12)


#this can then be used in all the same plotting functions above.
plot.twisst.summary(chr16_top12, lwd=3, cex=0.7)

pdf("results/2021-08-07-twisst/summary_top12_chr16.pdf", width=31, height=8)
plot.twisst.summary(chr16_top12, lwd=3, cex=0.7)
dev.off()

```

These four topologies represent the following evolutionary scenarios:

* topo116: Introgression from *S. invicta* to *S. richteri*
* topo119: Trans-species polymorphism
* topo111: Introgression from *S. richteri* to *S. invicta*
* topo118: Species tree topology (or two parallel origins of Sb)

To make a more clear view of this, we decided to make a barplot where we plot the number of trees supporting a given tree.

```{r}

# Get the topologies of interest
top_topologies <- rev(c("topo116", "topo641", "topo111", "topo119", "topo118"))

top_topologies_names <- rev(c(
  topo116 = "Introgression from S. invicta to S. richteri",
  topo641 = "Introgression from S. invicta to S. richteri (discordant early branches)",
  topo111 = "Introgression from S. richteri to S. invicta",
  topo119 = "Trans-species polymorphism",
  topo118 = "Species tree",
  other   = "Other"
))

top_topologies_index <- match(top_topologies, names(all_twisst$topos))

# Get the best topologies:
# Top four
other_topos <- names(all_twisst$topos)[-top_topologies_index]

```


```{r}

# Make a table with the max support for window
top4_twisst <- subset.twisst.by.topos(all_twisst, top_topologies_index)
top4_twisst_w <- do.call(rbind, top4_twisst$weights)

# Others
other_twisst <- subset.twisst.by.topos(all_twisst, other_topos)
other_twisst_w <- do.call(rbind, other_twisst$weights)

stopifnot(!colnames(top4_twisst_w) %in% colnames(other_twisst_w))

# Coordinates
# top4_twisst_w <- cbind(
#   full_coords[,1:4],
#   top4_twisst_w,
#   Other = apply(other_twisst_w, 1, sum)
# )

top4_twisst_w <- cbind(
  full_coords[,1:5],
  top4_twisst_w,
  other_twisst_w
)

library(ggthemes)

top4_twisst_w %>%
  arrange(chrom, start, end) %>%
  mutate(ID = row_number()) %>%
  pivot_longer(cols = !c(ID, chrom, start, end, mid, region),
               names_to = "Topology_code",
               values_to = "Relative_weight") %>%
  mutate(Topology = recode(Topology_code,
                           topo116 = "Introgression (S. invicta to S. richteri)",
                           topo119 = "Trans-species polymorphism",
                           topo111 = "Introgression (S. richteri to S. invicta)",
                           topo118 = "Species tree")) %>%
  mutate(Topology_type = recode(Topology_code,
                           topo116 = "Introgression",
                           topo119 = "Trans-species\npolymorphism",
                           topo111 = "Introgression",
                           topo641 = "Introgression",
                           topo118 = "Species tree",
                           topo643 = "Species tree",
                           topo13  = "Species tree")) -> weight_by_window

weight_by_window$Topology_type[!weight_by_window$Topology_type %in% c("Introgression",
                                                     "Trans-species\npolymorphism",
                                                     "Species tree")] <- "Other"

# Subset dataset to best
weight_by_window %>%
  group_by(ID, chrom, start, end, mid, region) %>%
  summarise(Best_topology      = Topology[which.max(Relative_weight)],
            Best_topology_type = Topology_type[which.max(Relative_weight)],
            Relative_weight    = max(Relative_weight)) %>%
  mutate(region = gsub("chr16[A-Z]", "chr16\n(outside supergene)", region)) %>%
  mutate(region = gsub("chr16_supergene", "chr16\n(supergene)", region)) ->
  best_topologies_per_window

# Subset to windoes where best topology has a weight of over 0.9
best_topologies_per_window %>%
  filter(Relative_weight > 0.9) %>%
  mutate(Best_topology = as.factor(Best_topology)) %>%
  mutate(Best_topology_type = as.factor(Best_topology_type)) %>%
  mutate(Best_topology = fct_relevel(Best_topology,
             "topo641",
             "Introgression (S. richteri to S. invicta)",
             "Introgression (S. invicta to S. richteri)",
             "topo643",
             "topo13",
             "Trans-species polymorphism",
             "Species tree")) ->
  best_topologies_per_window_filtered

```

```{r}

type_order <- c("Introgression",
               "Trans-species\npolymorphism",
               "Species tree",
               "Other")

topology_order <- c("Introgression (S. invicta to S. richteri)",
                     "Introgression (S. richteri to S. invicta)",
                     "Trans-species polymorphism",
                     "Species tree",
                     "topo113",
                     "topo641",
                     "topo643",
                     "topo13")

my_colours <-  c("Introgression (S. invicta to S. richteri)" = "#0072B2",
                 "Introgression (S. richteri to S. invicta)" = "#56B4E9",
                 "Trans-species polymorphism" = "#F0E442",
                 "Species tree" = "#009E73", 
                 "topo113" = "#E69F00",
                 "topo641" = "#D55E00",
                 "topo643" = "#CC79A7",
                 "topo13"  = "#000000")


topology_fill <- scale_fill_manual(
  name = "Best topology",
  breaks = names(my_colours),
  values = my_colours
)


best_topologies_per_window_filtered %>%
  group_by(region, Best_topology_type, Best_topology) %>%
  count %>%
  group_by(region) %>%
  mutate(perc = n / sum(n)) %>%
  ggplot(aes(x = Best_topology_type, fill = Best_topology, y = perc)) +
    geom_bar(stat = "identity") + 
    scale_y_continuous(labels=scales::percent, name = "Windows supporting topology") +
    facet_grid(rows=vars(region)) +
    theme_bw() +
    scale_x_discrete(name = NULL, limits = type_order) +
    topology_fill +
    theme(legend.position = c(0.25, 0.7),
          legend.background = element_rect(fill="white",
                                           size=0.5,
                                           linetype="solid", 
                                           colour ="darkgrey")) -> p

ggsave(p, file = "results/twisst_figures/best_topology_per_window.pdf",
       width = 7, height = 4)
p

```

```{r}

best_topologies_per_window_filtered %>%
  group_by(region, Best_topology_type, Best_topology) %>%
  count %>%
  group_by(region) %>%
  mutate(perc = n / sum(n)) %>%
  ggplot(aes(x = Best_topology_type, fill = Best_topology, y = perc)) +
    geom_bar(stat = "identity") + 
    facet_grid(cols=vars(region)) +
    scale_y_continuous(labels=scales::percent, name = "Windows supporting topology") +
    theme_bw() +
    scale_x_discrete(name = NULL, limits = type_order) +
    theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1)) +
    topology_fill +
    theme(legend.position = c(0.25, 0.7),
          legend.background = element_rect(fill="white",
                                           size=0.5,
                                           linetype="solid", 
                                           colour ="darkgrey")) -> p

ggsave(p, file = "results/twisst_figures/best_topology_per_window_facet.pdf",
        width = 7, height = 4)
p


```

We need to make a table with the counts:

```{r}

best_topologies_per_window_filtered %>%
  group_by(region) %>%
  count -> a

a

```
There are a total of `r sum(a$n)` windows with relative weight greater than 0.9, out of `r nrow(best_topologies_per_window)` windows (or `r 100* sum(a$n)/nrow(best_topologies_per_window)`%).

```{r}

best_topologies_per_window_filtered %>%
  group_by(region, Best_topology_type) %>%
  count %>%
  group_by(region) %>%
  mutate(perc = round(100* n / sum(n)))

best_topologies_per_window_filtered %>%
  group_by(region, Best_topology_type, Best_topology) %>%
  count %>%
  group_by(region) %>%
  mutate(perc = round(100* n / sum(n)))

```

We also need to plot the trees

```{r, fig.width = 6, fig.height = 4}

topology_order_for_trees <- c("topo116",
                              "topo111",
                              "topo119",
                              "topo118",
                              "topo113",
                              "topo641",
                              "topo643",
                              "topo13")

my_colours <-  c("Introgression (S. invicta to S. richteri)" = "#0072B2",
                 "Introgression (S. richteri to S. invicta)" = "#56B4E9",
                 "Trans-species polymorphism" = "#F0E442",
                 "Species tree" = "#009E73", 
                 "topo113" = "#E69F00",
                 "topo641" = "#D55E00",
                 "topo643" = "#CC79A7",
                 "topo13"  = "#000000")

pdf("results/twisst_figures/topology_per_type.pdf",
    width = 6, height = 3)

par(mfrow=c(3,4),
    mar= c(0,5,0,0),
    mai = c(0.05, 0, 0.05, 0))

for (i in 1:length(topology_order)) {
  plot(all_twisst$topos[[topology_order_for_trees[i]]],
     type       ="cladogram",
     edge.color = my_colours[i],
     edge.width = 2,
     label.offset = 0.5)
  text(topology_order_for_trees[i], x = 2, y = 1)
}

dev.off()

```

## Chromosome level support

```{r}

# inversions
inversions <- data.frame(
  Inversion = c("In(16)1", "In(16)2", "In(16)3"),
  start     = c(17621722 + 351197, 16509882, 11732910),
  end       = c(27455557 + 694331, 17621722 + 340966, 15842782 + 635328),
  region    = "chr16\n(supergene)",
  chrom = "chr16"
)

inversions


lm <- read.csv("results/linkage_map_with_inversion.csv", header = TRUE)
colnames(lm)[1] <- "chrom"

lm <- filter(lm, chrom %in% c("chr1", "chr16"))
lm$inversion[lm$inversion == "no_known_inversion"] <- "No known inversion"
```

```{r best_topology, fig.width=12, fig.height = 8}

topology_order <- c("Introgression (S. invicta to S. richteri)",
                     "Introgression (S. richteri to S. invicta)",
                     "Trans-species polymorphism",
                     "Species tree",
                     "Other")
# best_topologies_per_window_plot <- best_topologies_per_window
# 
# best_topologies_per_window_plot$Best_topology[!best_topologies_per_window_plot$Best_topology %in%
#                                                 topology_order] <- "Other"
#   

my_colours <-  c("Introgression (S. invicta to S. richteri)" = "#0072B2",
                 "Introgression (S. richteri to S. invicta)" = "#56B4E9",
                 "Trans-species polymorphism" = "#F0E442",
                 "Species tree" = "#009E73", 
                 "Other" = "#E69F00")

scaffold_fill <- c(
  "In(16)1" = "purple",
  "In(16)2" = "light blue",
  "In(16)3" = "blue",
  "No known inversion" = "grey"
)

scaffold_fill_breaks <- c(
  "In(16)1",
  "In(16)2",
  "In(16)3",
  "No known inversion"
)


best_topologies_per_window %>%
  mutate(Topology_type = recode(Best_topology,
                             topo116 = "Introgression (S. invicta to S. richteri)",
                             topo119 = "Trans-species polymorphism",
                             topo111 = "Introgression (S. richteri to S. invicta)",
                             topo116 = "Introgression (S. invicta to S. richteri)",
                             topo118 = "Species tree",
                             topo643 = "Species tree",
                             topo13  = "Species tree")) -> best_topologies_per_window_plot

best_topologies_per_window_plot$Topology_type[!best_topologies_per_window_plot$Topology_type %in%
                                                 topology_order] <- "Other"

best_topologies_per_window_plot %>%
  mutate(mid = start + (end - start)/2) -> best_topologies_per_window_plot

ggplot() +
    geom_rect(data = lm, aes(xmin = chr_start, xmax = chr_end,
                                     ymin = 0.05, ymax = 0.15, fill = inversion),
              alpha = 0.75) +
     geom_point(data = best_topologies_per_window_plot,
               aes(x      = mid,
               y      = Relative_weight,
               colour = Topology_type,
               shape = Topology_type)) +
    facet_grid(rows=vars(chrom)) +
    theme_bw() +
    scale_colour_manual(name = "Best topology",
                        breaks = names(my_colours),
                        values = my_colours) +
    scale_shape_manual(name = "Best topology",
                       breaks = names(my_colours),
                       values = 15:19) +
    scale_fill_manual(name   = "Inversion",
                      breaks = scaffold_fill_breaks,
                      values = scaffold_fill) +
    ylab("Relative weight of best topology") +
    xlab("Window midpoint coordinate") +
  geom_hline(yintercept = 0.9, colour = "grey") +
  scale_x_continuous(labels = scales::label_number(scale = 1/1e6,
                                                   suffix = " Mb")) -> p


p +  theme(legend.position="top",
        legend.direction='vertical') -> p

ggsave(p, file="results/twisst_figures/chromosome.pdf", width=8  , height = 6)

p

```

```{r , fig.width = 14}

top_topologies <- rev(c("topo116", "topo641", "topo111", "topo119", "topo118"))

top_topologies_names <- rev(c(
  topo116 = "Introgression from S. invicta to S. richteri",
  topo641 = "Introgression from S. invicta to S. richteri (discordant early branches)",
  topo111 = "Introgression from S. richteri to S. invicta",
  topo119 = "Trans-species polymorphism",
  topo118 = "Species tree",
  other   = "Other"
))

topology_colour <- colorblind_pal()(length(top_topologies_names) + 1)[-1]
topology_colour <- c(topology_colour[1], topology_colour[6], topology_colour[2],
                     topology_colour[3], topology_colour[4], topology_colour[5])
topology_colour <- rev(topology_colour)
  
# Sum the relative weights of all "Other" topologies
weight_by_window %>%
  select(-Topology_type, -Topology) %>%
  pivot_wider(names_from = Topology_code, values_from = Relative_weight) %>%
  rowwise(c(chrom:ID, top_topologies)) %>%
  summarise(other = sum(c_across(contains("topo")))) %>%
  pivot_longer(cols = c(contains("topo"), other),
               names_to = "Topology_id",
               values_to = "Relative_weight") %>%
  mutate(Topology_id = as.factor(Topology_id)) %>%
  mutate(Topology_id = fct_relevel(Topology_id, c("other", top_topologies))) -> weight_by_window_with_other

# Make cumulative sums
weight_by_window_with_other %>%
  group_by(ID) %>%
  arrange(Topology_id, by_group = TRUE) %>%
  mutate(ymax = cumsum(Relative_weight)) %>%
  mutate(ymin = ymax - Relative_weight) %>%
  ungroup -> weight_by_window_with_other_rect

```


```{r fig.width = 14, fig.height = 4}
weight_by_window_with_other_rect %>%
  ggplot() +
  geom_rect(aes(xmin = start,
                xmax = end,
                ymin = ymin,
                ymax = ymax,
                fill = Topology_id)) +
  facet_grid(rows=vars(chrom)) +
  theme_bw() +
  scale_fill_manual(breaks = rev(names(top_topologies_names)),
                    labels = rev(top_topologies_names),
                    values = rev(topology_colour)) +
  ylab("Relative weight of topology") +
  xlab("Window position in chromosome") +
  scale_x_continuous(labels = scales::label_number(scale = 1/1e6,
                                                   suffix = " Mb")) -> p

p
```

```{r fig.width = 14, fig.height = 4}
weight_by_window_with_other_rect %>%
  ggplot() +
  geom_line(aes(x = mid,
                y = Relative_weight,
                group = Topology_id,
                colour = Topology_id)) +
  theme_bw() +
  scale_colour_manual(breaks = rev(names(top_topologies_names)),
                    labels = rev(top_topologies_names),
                    values = rev(topology_colour)) +
    facet_grid(rows=vars(chrom)) +
  ylab("Relative weight of topology") +
  xlab("Window middle coordinate in chromosome 16")

```

```{r fig.width = 14, fig.height = 4}

# p <- ggplot(data=weight_by_window_with_other_rect,
#             aes(x=mid, group=Topology_id, fill=Topology_id)) +
#     geom_density(adjust=1.5, position="fill") +
#   theme_bw() +
#     facet_grid(rows=vars(chrom)) +
#   scale_fill_manual(breaks = rev(names(top_topologies_names)),
#                     labels = rev(top_topologies_names),
#                     values = rev(topology_colour)) +
#   ylab("Relative weight of topology") +
#   xlab("Window middle coordinate in chromosome 16")
# p
# 


weight_by_window_with_other_rect %>%
  mutate(Topology_id = fct_relevel(Topology_id, rev(levels(weight_by_window_with_other_rect$Topology_id)))) %>%
  ggplot(aes(x = mid,
             y = Relative_weight,
             fill = Topology_id)) +
    geom_area() +
  theme_bw() +
  scale_fill_manual(breaks = rev(names(top_topologies_names)),
                    labels = rev(top_topologies_names),
                    values = rev(topology_colour)) +
    facet_grid(rows=vars(chrom)) +
  ylab("Relative weight of topology") +
  xlab("Window middle coordinate in chromosome 16")

```

## Windows per inversion

```{r, echo=TRUE}

lm %>%
  filter(inversion == "In(16)3") %>%
  arrange(chr_start) -> inv_3_reg

as.data.frame(inv_3_reg)

best_topologies_per_window %>%
  filter(chrom == "chr16", start > 11732910, end < 16339881) -> inv_3_windows

stopifnot(nrow(inv_3_windows) == 0)

inv_3_size <- sum(inv_3_reg$chr_end - inv_3_reg$chr_start)

```

The inversion In(16)3 is `r round(inv_3_size/1e6, 2)` Mbp long. There were no windows with a midpoint within its mapped range.

```{r, echo=TRUE}

lm %>%
  filter(inversion == "In(16)2") -> inv_2_reg

as.data.frame(inv_2_reg)

best_topologies_per_window %>%
  filter(chrom == "chr16", mid > 16509882, mid < 17962688) -> inv_2_windows

inv_2_windows_high <- inv_2_windows[inv_2_windows$Relative_weight > 0.9,]

stopifnot(inv_2_windows_high$Best_topology_type == "Introgression")

inv_2_size <- sum(inv_2_reg$chr_end - inv_2_reg$chr_start)

```

The inversion In(16)2 region of the supergene is `r round(inv_2_size/1e6, 2)` Mbp long. There were `r nrow(inv_2_windows)` windows in In(16)2 region of the supergene. Out of these, `r nrow(inv_2_windows_high)` had a single topology with a weight greater than 0.9, both supporting the _S. invicta_ to _S. richteri_ introgression scenario.

```{r}

inv_2_windows %>%
  as.data.frame()

```

```{r, echo=TRUE}

lm %>%
  filter(inversion == "In(16)1") %>%
  arrange(chr_start) -> inv_1_reg

inv_1_reg %>%
  slice(c(1,n())) %>%
  as.data.frame()

best_topologies_per_window %>%
  filter(chrom == "chr16", mid > 17972920, mid < 28149888) -> inv_1_windows

inv_1_windows_high <- inv_1_windows[inv_1_windows$Relative_weight > 0.9,]

inv_1_windows_high_int  <- sum(inv_1_windows_high$Best_topology == "Introgression (S. invicta to S. richteri)")
inv_1_windows_high_anc  <- sum(inv_1_windows_high$Best_topology == "Trans-species polymorphism")

inv_1_size <- sum(inv_1_reg$chr_end - inv_1_reg$chr_start)

```

The inversion In(16)1 region of the supergene is `r round(inv_1_size/1e6, 2)` Mbp long. There were `r nrow(inv_1_windows)` windows in In(16)1 region of the supergene. Out of these, `r nrow(inv_1_windows_high)` (`r round(100 * nrow(inv_1_windows_high)/nrow(inv_1_windows))`%) had a single topology with a relative weight greater than 0.9. Of these, `r inv_1_windows_high_int` (`r round(100*inv_1_windows_high_int/nrow(inv_1_windows_high))`%) suported the _S. invicta_ to _S. richteri_ introgression scenario, while `r inv_1_windows_high_anc` (`r round(100*inv_1_windows_high_anc/nrow(inv_1_windows_high))`%) supported the next best topology, consistent with the Trans-species ancestral polymorphism.

```{r}

best_topologies_per_window$Inversion <- "No known inversion"
  
best_topologies_per_window$Inversion[best_topologies_per_window$chrom == "chr16" &
                             best_topologies_per_window$mid > 11732910 &
                             best_topologies_per_window$mid < 16339881 ] <- "In(16)3"

best_topologies_per_window$Inversion[best_topologies_per_window$chrom == "chr16" &
                             best_topologies_per_window$mid > 16509882 &
                             best_topologies_per_window$mid < 17962688 ] <- "In(16)2"

best_topologies_per_window$Inversion[best_topologies_per_window$chrom == "chr16" &
                             best_topologies_per_window$mid > 17972920 &
                             best_topologies_per_window$mid < 28149888 ] <- "In(16)1"

```

The number of windows with a single topology with a relative wight greater than 0.9 is:

```{r}

best_topologies_per_window %>%
  group_by(Inversion) %>%
  summarise(window_number = n(),
            high_weight_windows = sum(Relative_weight > 0.9),
            high_weight_windows_per = round(100 * sum(Relative_weight > 0.9)/n())) %>%
  as.data.frame()

```

The number of windows with a single topology with a relative wight greater than 0.9 is:

```{r}

best_topologies_per_window %>%
  filter(Relative_weight > 0.9) %>%
  group_by(Inversion, Best_topology) %>%
  summarise(window_number = n()) %>%
  group_by(Inversion) %>%
  mutate(window_per = round(100 * window_number / sum(window_number), 1)) %>%
  as.data.frame()

```
