---
title: "Comparing ILS and introgression scenarios using SNPs"
author: "Rodrigo Pracana, QMUL"
date: "04/10/2021"
output: html_document
---

```{r setup, include=FALSE, cache = FALSE}


make_sci <- function(x) {
    sprintf("%.1fx10^{%d}",
          x/10^floor(log10(abs(x))),
          floor(log10(abs(x))))
}

pretty_number_inner <- function(x) {
  if (is.na(x)) {
    x <- NA
  } else if (x == 0) {
    x <- as.character(x)
  } else if (x < 1e-4) {
    x <- as.character(make_sci(x))
  } else if  (x < 1e-2) {
     x <- as.character(round(x, 4))
  } else if (x < 1e4) {
     x <- as.character(round(x, 2))
  } else {
     x <- as.character(make_sci(x))
  }
  return(x)
}

pretty_numbers <- function(x) {
  sapply(x, pretty_number_inner)
}

knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)


library(tidyverse)
library(patchwork)
library(broom)

dir.create(recursive = T, "results/fig")

```

```{r}

read_dfoil <- function(path, species) {
  dfoil <- read.table(path, comment.char = '&', header = T)
  dfoil$species <- species

  dfoil %>%
    rename(gene = X.chrom) %>%
    mutate(gene = gsub(".*\\/", "", gene)) %>%
    mutate(gene = gsub("\\.fa", "", gene)) %>%
    return()
}

dfoil_to_long <- function(dfoil) {
  dfoil %>%
    select(gene, coord, species, ends_with("_stat"), ends_with("_Pvalue")) %>%
    pivot_longer(ends_with(c("_stat","_Pvalue")), names_to = "D_type", values_to = "D_value") %>%
    separate(D_type, c("D_type", "stat")) %>%
    pivot_wider(names_from = stat, values_from = D_value) %>%
    rename(D_value = stat, P_value = Pvalue) %>%
    mutate(D_type = fct_relevel(D_type, "DFO", "DIL", "DFI", "DOL")) %>%
    return()
}

meg  <- read_dfoil("results/dfoil_output/megergates.dfoil", "megergates")
int  <- read_dfoil("results/dfoil_output/interrupta.dfoil", "interrupta")
adrx <- read_dfoil("results/dfoil_output/xadr.dfoil", "AdRX")
rich <- read_dfoil("results/dfoil_output/richteri.dfoil", "richteri")

all_dfoil <- rbind(meg, int, adrx, rich)

all_long <- dfoil_to_long(all_dfoil)

```

```{r, eval=FALSE}

dfoil_numbers_to_long <- function(dfoil) {
  dfoil %>%
    select(gene, coord, species, ends_with("_left"), ends_with("_right")) %>%
    pivot_longer(c(ends_with("_left"), ends_with("_right"))) %>%
    separate(name, c("D_type", "hand")) %>%
    pivot_wider(names_from = "hand", values_from = "value") %>%
    mutate(D_type = fct_relevel(D_type, "DFO", "DIL", "DFI", "DOL")) %>%
    return()
}

all_dfoil_snps <- dfoil_numbers_to_long(all_dfoil)

ggplot(all_dfoil_snps) +
  geom_point(aes(x = right, y = left), alpha = 0.5) +
    facet_grid(rows = vars(species), cols = vars(D_type)) +
    geom_abline(intercept = 0, slope = 1) +
    theme_bw() +
    xlim(0, 100) + ylim(0,100)


```

## Introduction


Our main aim is to use alleles combination patterns (analagous to the ABBA-BABA test) to tell apart between the three hypothesis for the origin of the supergene.

Taking the comparison between _S. invicta_ and _S. richteri_ as an example, the parallel origin hypothesis has following tree topology, mirroring the species tree:

((richteri_B, richteri_b),(invicta_b, invicta_B)),saevissima).

The ancestral polymorphism hypothesis follows the following tree:

((richteri_b, invicta_b),(invicta_B,richteri_B)),saevissima)

The introgression hypothesis follows one of the following trees:

((richteri_b, invicta_b),invicta_B),richteri_B)),saevissima)

((richteri_b, invicta_b),richteri_B),invicta_B)),saevissima)

We use the DFOIL approach charecterise allele distributions across the species. DFOIL considers symmetrical five-taxon the "species tree" topology, with the introgressed species on the left two branches, _S. invicta_ in the two middle branches and the outgroup _S. saevissima_ on the right branch. As an example, the _S. richeteri_ tree we considered is:

((richteri_B, richteri_b),(invicta_b, invicta_B)),saevissima).

We chose a single sample to represent each species. For each gene in the supergene, we counted the occurance of every possible allele combination. We repeated this for each four species with a putative introgression of Sb.

```{r}

read_allele_counts <- function(path, species) {
  dfoil <- read.table(path, comment.char = '&', header = T)
  dfoil$species <- species

  dfoil %>%
    rename(gene = X.chrom) %>%
    mutate(gene = gsub(".*\\/", "", gene)) %>%
    mutate(gene = gsub("\\.fa", "", gene)) %>%
    relocate(species, .after=position) %>%
    return()
}

richteri_counts   <- read_allele_counts("results/allele_counts/richteri.counts", "richteri")
megergates_counts <- read_allele_counts("results/allele_counts/megergates.counts", "megergates")
interrupta_counts <- read_allele_counts("results/allele_counts/interrupta.counts", "interrupta")
adrx_counts       <- read_allele_counts("results/allele_counts/adrx.counts", "AdRX")

all_counts <- rbind(richteri_counts, megergates_counts, interrupta_counts, adrx_counts)

```

The mean count for gene allele combination (averages across genes) is shown on this table.

```{r}

all_counts %>%
  pivot_longer(!c(gene, position, species)) %>%
  group_by(name, species) %>%
  summarise(mean_count = round(mean(value),1)) %>%
  pivot_wider(names_from = species, values_from = mean_count)

```

## Evidence that Sb introgressed from _S. invicta_ to the remaining species

We tested the hypothesis that the SB alleles of the species in which Sb evolved will have more private alleles in common with Sb than the other species. Most genes had a small number of such alleles. Nevertheless, on average _S. invicta_ had more private alleles in common with Sb than each of the focal species (two-tail paired t-test, Bonferonni-corrected p < 0.05).

```{r, fig.height = 6, fig.width = 6}

# Alleles shared between Sb and invicta SB
# ((richteri_B, richteri_b),(invicta_b, invicta_B)),saevissima).
# ABBBA
# AABBA
# ABABA
# Alleles shared between Sb and richteri SB
# ((richteri_B, richteri_b),(invicta_b, invicta_B)),saevissima).
# BBBAA
# BABAA
# BBAAA

all_counts %>%
  mutate(shared_lb_invicta_bb = ABBBA,
         shared_lb_focal_bb = BBBAA) %>%
  mutate(species = as.factor(paste0("S. ", species))) ->
  shared_lb_each_bb

shared_lb_each_bb %>%
  group_by(species) %>%
  nest() %>%
  mutate(model        = map(data,
                            ~ t.test(.$shared_lb_invicta_bb,
                                          .$shared_lb_focal_bb,
                                          paired = TRUE)),
         results      = map(model, tidy),
         mean_invicta = map(data, ~ mean(.$shared_lb_invicta_bb)),
         mean_focal   = map(data, ~ mean(.$shared_lb_focal_bb)),
         sum_invicta  = map(data, ~ sum(.$shared_lb_invicta_bb)),
         sum_focal    = map(data, ~ sum(.$shared_lb_focal_bb))) %>%
  unnest(c(results, mean_invicta, mean_focal, sum_invicta, sum_focal)) %>%
  select(-data, -model, -method, -alternative) %>%
  mutate(perc_diff = round(100 * (mean_invicta - mean_focal) / mean_focal),
         corrected_p = 4 * p.value) ->
  shared_lb_each_bb_summary

as.data.frame(shared_lb_each_bb_summary)


```


```{r}

# Add P-value to plot for printing
 parse_p <- function(p_value, facetting_var) {

  if (is.na(p_value)) {
    p_value <- paste0("atop('", expr(!!facetting_var), "',",
                     ")")
  } else if (p_value >= 0.05) {
    p_value <- paste0("atop('", expr(!!facetting_var), "',",
                     "'corrected p =", expr(!!round(p_value, 2)),"')")
  } else if (p_value < 0.05 & p_value >= 0.001) {
    p_value <- paste0("atop('", expr(!!facetting_var), "',",
                     "'corrected p = ", expr(!!round(p_value, 3)),"')")
  } else {
    p_value <- paste0("atop('", expr(!!facetting_var), "',",
                     "'corrected p < 10'", "^", expr(!!(floor(log10(abs(p_value))) + 1)),")")
  }
  return(p_value)
}

species_levels <- levels(shared_lb_each_bb$species)
p_vector      <- shared_lb_each_bb_summary$corrected_p[match(species_levels, shared_lb_each_bb_summary$species)]

# Make P value expression
p_vector <- sapply(1:length(p_vector), function(i) parse_p(p_vector[i],
                                                           species_levels[i]))

shared_lb_each_bb$species <- factor(shared_lb_each_bb$species,
                                    labels = p_vector)

shared_lb_each_bb %>%
  ggplot() +
  geom_point(aes(x = shared_lb_invicta_bb,
                 y =  shared_lb_focal_bb), alpha = 0.5) +
  facet_wrap(vars(species), ncol = 2, labeller = labeller(species = label_parsed)) +
  theme_bw() +
  geom_abline(slope=1, intercept=0) +
  coord_fixed(xlim = c(0, 21.5), ylim = c(0, 21.5)) +
  xlab("Shared between Sb and the S. invicta SB") +
  ylab("Shared between Sb and the focal SB") -> p2

p2

ggsave(p2, file = "results/lb_shared_each_bb.pdf", width = 6, height = 6)

```

## Conclusion

The fact that SB of _S. invicta_ shared more alleles with the Sb branch than the SB of any other species is consistent with an introgression from _S. invicta_ to each of the other species.
