---
title: "Applying the dfoil method to study the introgression of Sb"
author: "Rodrigo Pracana, QMUL"
date: "21/10/2021"
output: html_document
---

```{r setup, include=FALSE, cache = FALSE}


make_sci <- function(x) {
    sprintf("%.1fx10^{%d}",
          x/10^floor(log10(abs(x))),
          floor(log10(abs(x))))
}

pretty_number_inner <- function(x) {
  if (is.na(x)) {
    x <- NA
  } else if (x == 0) {
    x <- as.character(x)
  } else if (x < 1e-4) {
    x <- as.character(make_sci(x))
  } else if  (x < 1e-2) {
     x <- as.character(round(x, 4))
  } else if (x < 1e4) {
     x <- as.character(round(x, 2))
  } else {
     x <- as.character(make_sci(x))
  }
  return(x)
}

pretty_numbers <- function(x) {
  sapply(x, pretty_number_inner)
}

knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)


library(tidyverse)
library(patchwork)
library(broom)

dir.create(recursive = T, "results/fig")

```

```{r}

read_dfoil <- function(path, species) {
  dfoil <- read.table(path, comment.char = '&', header = T)
  dfoil$species <- species

  dfoil %>%
    rename(gene = X.chrom) %>%
    mutate(gene = gsub(".*\\/", "", gene)) %>%
    mutate(gene = gsub("\\.fa", "", gene)) %>%
    return()
}

dfoil_to_long <- function(dfoil) {
  dfoil %>%
    select(gene, coord, species, ends_with("_stat"), ends_with("_Pvalue")) %>%
    pivot_longer(ends_with(c("_stat","_Pvalue")), names_to = "D_type", values_to = "D_value") %>%
    separate(D_type, c("D_type", "stat")) %>%
    pivot_wider(names_from = stat, values_from = D_value) %>%
    rename(D_value = stat, P_value = Pvalue) %>%
    mutate(D_type = fct_relevel(D_type, "DFO", "DIL", "DFI", "DOL")) %>%
    return()
}

meg  <- read_dfoil("results/dfoil_output/megergates.dfoil", "megergates")
int  <- read_dfoil("results/dfoil_output/interrupta.dfoil", "interrupta")
adrx <- read_dfoil("results/dfoil_output/xadr.dfoil", "AdRX")
rich <- read_dfoil("results/dfoil_output/richteri.dfoil", "richteri")

all_dfoil <- rbind(meg, int, adrx, rich)

all_long <- dfoil_to_long(all_dfoil)

```

```{r, eval=FALSE}

dfoil_numbers_to_long <- function(dfoil) {
  dfoil %>%
    select(gene, coord, species, ends_with("_left"), ends_with("_right")) %>%
    pivot_longer(c(ends_with("_left"), ends_with("_right"))) %>%
    separate(name, c("D_type", "hand")) %>%
    pivot_wider(names_from = "hand", values_from = "value") %>%
    mutate(D_type = fct_relevel(D_type, "DFO", "DIL", "DFI", "DOL")) %>%
    return()
}

all_dfoil_snps <- dfoil_numbers_to_long(all_dfoil)

ggplot(all_dfoil_snps) +
  geom_point(aes(x = right, y = left), alpha = 0.5) +
    facet_grid(rows = vars(species), cols = vars(D_type)) +
    geom_abline(intercept = 0, slope = 1) +
    theme_bw() +
    xlim(0, 100) + ylim(0,100)


```

## Introduction

Our main aim is to use alleles combination patterns (analagous to the ABBA-BABA test) to tell apart between the three hypothesis for the origin of the supergene.

Taking the comparison between _S. invicta_ and _S. richteri_ as an example, the parallel origin hypothesis has following tree topology, mirroring the species tree:

((richteri_B, richteri_b),(invicta_b, invicta_B)),saevissima).

The ancestral polymorphism hypothesis follows the following tree:

((richteri_b, invicta_b),(invicta_B,richteri_B)),saevissima)

The introgression hypothesis follows one of the following trees:

((richteri_b, invicta_b),invicta_B),richteri_B)),saevissima)

((richteri_b, invicta_b),richteri_B),invicta_B)),saevissima)

We use the DFOIL approach charecterise allele distributions across the species. DFOIL considers symmetrical five-taxon the "species tree" topology, with the introgressed species on the left two branches, _S. invicta_ in the two middle branches and the outgroup _S. saevissima_ on the right branch. As an example, the _S. richeteri_ tree we considered is:

((richteri_B, richteri_b),(invicta_b, invicta_B)),saevissima).

We chose a single sample to represent each species. For each gene in the supergene, we counted the occurance of every possible allele combination. We repeated this for each four species with a putative introgression of Sb.

```{r}

read_allele_counts <- function(path, species) {
  dfoil <- read.table(path, comment.char = '&', header = T)
  dfoil$species <- species

  dfoil %>%
    rename(gene = X.chrom) %>%
    mutate(gene = gsub(".*\\/", "", gene)) %>%
    mutate(gene = gsub("\\.fa", "", gene)) %>%
    relocate(species, .after=position) %>%
    return()
}

richteri_counts   <- read_allele_counts("results/allele_counts/richteri.counts", "richteri")
megergates_counts <- read_allele_counts("results/allele_counts/megergates.counts", "megergates")
interrupta_counts <- read_allele_counts("results/allele_counts/interrupta.counts", "interrupta")
adrx_counts       <- read_allele_counts("results/allele_counts/adrx.counts", "AdRX")

all_counts <- rbind(richteri_counts, megergates_counts, interrupta_counts, adrx_counts)

```

The mean count for gene allele combination (averages across genes) is shown on this table.

```{r}

all_counts %>%
  pivot_longer(!c(gene, position, species)) %>%
  group_by(name, species) %>%
  summarise(mean_count = round(mean(value),1)) %>%
  pivot_wider(names_from = species, values_from = mean_count)

```

## No support for the parallel origin hypothesis

Under the parallel origin hypothesis, alleles would be shared between species only by ILS, meaning that a supergene variant in a species would be equally distant to both supergene variants in the other species.

The DFOIL approach gives us four measures. We expect all measures to be zero.

DFO tests whether the first taxon (SB in the focal species) is closer to the third (invicta Sb) or fourth (invicta SB) taxon. For all species, we get a negative value on average, indicating that focal SB variant is closer to the SB variant of _S. invicta_ than the Sb variant of _S. invicta_.

DIL measures whether the second taxon (focal Sb) is closer to the third (invicta Sb) or the fourth (invicta SB). We have a positive value on average for all species, indicating that Sb variants share more alleles than expected under ILS.

DFI measures whether the third taxon (invicta Sb) is closer to the first (focal SB) or the second (focal Sb). Wehave a negative value on average for all focal species, again indicating that Sb variants share more alleles than expected under ILS.

DOL measures whether the fourth taxon (invicta SB) is closer to the first (focal SB) or second (focal Sb). For all species except _S. AdRX_ , we get a slightly positive value, indicating that the SB variant of _S. invicta_ is closer to the SB variant of the focal species than to the Sb variant of the focal species.

```{r, fig.height = 7, fig.width = 10}

all_long %>%
  mutate(significant = ifelse(P_value < 0.05, "p < 0.05", "p > 0.05")) %>%
  ggplot(aes(x = D_type, y = D_value)) +
    # geom_jitter(aes(colour = significant)) +
    geom_boxplot(width=0.5, fill ="transparent") +
    geom_hline(yintercept = 0) +
    facet_wrap(vars(species), ncol = 2) +
    theme_bw()

```

These patterns can be seen below, where I add the proportion of genes that have a significantly positive of negative D for each D type and for each species (chi-square goodness-of-fit test, uncorrected p < 0.05). I also plot the P-value distributions.

```{r}

all_long %>%
  mutate(significant = ifelse(P_value < 0.05, "p < 0.05", "p > 0.05")) %>%
  group_by(D_type, species) %>%
  summarise(mean_D = mean(D_value),
            proportion_positive = round(sum(P_value < 0.05 & D_value > 0) / n(), 2),
            proportion_negative = round(sum(P_value < 0.05 & D_value < 0) / n(), 2))

all_dfoil %>%
  select(species, gene, DFO_Pvalue, DIL_Pvalue, DFI_Pvalue, DOL_Pvalue) %>%
  pivot_longer(-c(species, gene)) %>%
  mutate(D = gsub("_.*", "", name)) %>%
  mutate(D = fct_relevel(D, "DFO", "DIL", "DFI", "DOL")) %>%
  ggplot() +
    geom_histogram(aes(x = value), colour = "black", fill = "white") +
    facet_grid(cols = vars(D), rows = vars(species)) +
    theme_bw()

```

From these patterns, it is clear that the "parallel origin" hypothesis, where alleles follow the species tree and ILS, is not correct.

At face value, the DFO and DOL values in our results could indicate that the SB have a more recent divergence than the separation between SB and Sb, which in turn would support the ancestral polymorphism hypothesis. However, both our alternative scenarios (ancestral polymorphism or introgression) are extremely discordant from the species tree. The measures of DFO and DOL are heavily skewed by alleles that are private to either SB (BAABA) or Sb (ABBAA), which could occur under either scenario.

We conclude that we cannot use dfoil to study the introgression of Sb between species.

## Appendix

### Alleles shared between SB and Sb

```{r, fig.height = 5, fig.width = 13}

ggplot(all_counts) +
  geom_point(aes(x = ABBBA, y =  BBBAA), alpha = 0.5) +
  facet_wrap(vars(species), ncol = 2) +
  theme_bw() +
  geom_abline(slope=1, intercept=0) +
  coord_fixed(xlim = c(0, 20), ylim = c(0, 20)) -> p1

ggplot(all_counts) +
  geom_point(aes(x = AABBA, y =  BABAA), alpha = 0.5) +
  facet_wrap(vars(species), ncol = 2) +
  theme_bw() +
  geom_abline(slope=1, intercept=0) +
  coord_fixed(xlim = c(0, 20), ylim = c(0, 20)) -> p2

ggplot(all_counts) +
  geom_point(aes(x = ABABA, y =  BBABA), alpha = 0.5) +
  facet_wrap(vars(species), ncol = 2) +
  theme_bw() +
  geom_abline(slope=1, intercept=0) +
  coord_fixed(xlim = c(0, 20), ylim = c(0, 20)) -> p3

p1 + p2 + p3

```

### Components for DIL

DIL is composed of the following four components. A right-skewed distribution supports the introgression hypothesis.

```{r, fig.width=12, fig.height = 6}

 ggplot(all_counts) +
  geom_histogram(aes(x = ABBAA - ABABA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) ->
  p1

ggplot(all_counts) +
  geom_histogram(aes(x = BBBAA - BBABA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) -> p2

ggplot(all_counts) +
  geom_histogram(aes(x = BAABA - BABAA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) ->
p3

ggplot(all_counts) +
  geom_histogram(aes(x = AAABA - AABAA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +  geom_vline(xintercept = 0) -> p4

p1 + p2 + p3 + p4 + plot_layout(ncol = 4)

```

### Components for DFI

DFI is composed of the following four components. A left-skewed distribution supports the introgression hypothesis.

```{r, fig.width=12, fig.height = 6}

 ggplot(all_counts) +
  geom_histogram(aes(x = BABAA - ABBAA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) ->
  p1

ggplot(all_counts) +
  geom_histogram(aes(x = BABBA - ABBBA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) -> p2

ggplot(all_counts) +
  geom_histogram(aes(x = ABABA - BAABA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) ->
p3

ggplot(all_counts) +
  geom_histogram(aes(x = ABAAA - BAAAA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +  geom_vline(xintercept = 0) -> p4

p1 + p2 + p3 + p4 + plot_layout(ncol = 4)

```

### Components for DFO

DFO is composed of the comparisons below. We see that the the negative skew is seen only for the first pannel and third pannel, which take into account the alleles that are private to each of the supergene variants (ABBAA for Sb and BAABA for SB).

```{r, fig.width=12, fig.height = 6}

 ggplot(all_counts) +
  geom_histogram(aes(x = BABAA - BAABA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) ->
  p1

ggplot(all_counts) +
  geom_histogram(aes(x = BBBAA - BBABA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) -> p2

ggplot(all_counts) +
  geom_histogram(aes(x = ABABA - ABBAA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) ->
p3

ggplot(all_counts) +
  geom_histogram(aes(x = AAABA - AABAA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +  geom_vline(xintercept = 0) -> p4

p1 + p2 + p3 + p4 + plot_layout(ncol = 4)

```

### Components for DOL

A similar pattern is seen for DOL, where the positive skew is also seen only for the first pannel and third pannel, relating to the alleles that are private to each of the supergene variants (ABBAA for Sb and BAABA for SB).

```{r, fig.width=12, fig.height = 6}

 ggplot(all_counts) +
  geom_histogram(aes(x = BAABA - ABABA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) ->
  p1

ggplot(all_counts) +
  geom_histogram(aes(x = BABBA - ABBBA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) -> p2

ggplot(all_counts) +
  geom_histogram(aes(x = ABBAA - BABAA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +
  geom_vline(xintercept = 0) ->
p3

ggplot(all_counts) +
  geom_histogram(aes(x = ABAAA - BAAAA), fill = "white", colour = "black") +
  facet_grid(rows = vars(species)) +
  theme_bw() +
  xlim(-35,35) +
  ylim(0,60) +  geom_vline(xintercept = 0) -> p4

p1 + p2 + p3 + p4 + plot_layout(ncol = 4)

```
